pkg/kallsyms/kallsyms.go

```diff
@@ -149,12 +151,53 @@ var (
 //
 //	kallsyms.SpecUpdateAddresses(spec, []string{"socket_file_ops"})
 func SpecUpdateAddresses(spec *ebpf.CollectionSpec, symbols []string) error {
-	kAllSymsFactory := NewKAllSyms
-	return specUpdateAddresses(kAllSymsFactory, spec, symbols)
+	return specUpdateAddresses(
+		[]symbolResolver{
+			symbolResolverFromKallsyms(NewKAllSyms),
+			symbolResolverFromEbpf(),
+		},
+		spec,
+		symbols,
+	)
+}
+
+func symbolResolverFromKallsyms(kAllSymsFactory func() (*KAllSyms, error)) symbolResolver {
```

Comment:
Why does it need a factory, can it receive `KAllSyms` directly?

Reply:
Because the unit test uses a different factory that does not requires root but provides its own alternative to `/proc/kallsyms`.

Reply:
What about something like:

```diff
diff --git pkg/kallsyms/kallsyms.go pkg/kallsyms/kallsyms.go
index 46522163..eb1d71b3 100644
--- pkg/kallsyms/kallsyms.go
+++ pkg/kallsyms/kallsyms.go
@@ -151,9 +151,14 @@ type symbolResolver func(symbol string) (uint64, error)
 //
 //     kallsyms.SpecUpdateAddresses(spec, []string{"socket_file_ops"})
 func SpecUpdateAddresses(spec *ebpf.CollectionSpec, symbols []string) error {
+       kAllSyms, err := NewKAllSyms()
+       if err != nil {
+               return err
+       }
+
        return specUpdateAddresses(
                []symbolResolver{
-                       symbolResolverFromKallsyms(NewKAllSyms),
+                       symbolResolverFromKallsyms(kAllSyms),
                        symbolResolverFromEbpf(),
                },
                spec,
@@ -161,16 +166,8 @@ func SpecUpdateAddresses(spec *ebpf.CollectionSpec, symbols []string) error {
        )
 }
 
-func symbolResolverFromKallsyms(kAllSymsFactory func() (*KAllSyms, error)) symbolResolver {
-       var kAllSyms *KAllSyms
+func symbolResolverFromKallsyms(kAllSyms *KAllSyms) symbolResolver {
        return func(symbol string) (uint64, error) {
-               if kAllSyms == nil {
-                       var err error
-                       kAllSyms, err = kAllSymsFactory()
-                       if err != nil {
-                               return 0, err
-                       }
-               }
                addr, ok := kAllSyms.symbolsMap[symbol]
                if !ok {
                        return 0, fmt.Errorf("symbol %q was not found in kallsyms", symbol)
diff --git pkg/kallsyms/kallsyms_test.go pkg/kallsyms/kallsyms_test.go
index 5274f334..d48f90c5 100644
--- pkg/kallsyms/kallsyms_test.go
+++ pkg/kallsyms/kallsyms_test.go
@@ -122,27 +122,27 @@ var specTest = &ebpf.CollectionSpec{
 func TestSpecRewriteConstants(t *testing.T) {
        t.Cleanup(resetState)
 
-       kAllSymsFactory := func() (*KAllSyms, error) {
-               kAllSymsStr := strings.Join([]string{
-                       "0000000000000000 A fixed_percpu_data",
-                       "ffffffffb4231f40 D bpf_prog_fops",
-                       "ffffffffb43723e0 d socket_file_ops",
-               }, "\n")
-
-               kAllSymsReader := strings.NewReader(kAllSymsStr)
-               kAllSyms, err := NewKAllSymsFromReader(kAllSymsReader)
-               require.Nil(t, err, "NewKAllSymsFromReader failed: %v", err)
-               return kAllSyms, nil
-       }
+       var err error
+
+       kAllSymsStr := strings.Join([]string{
+               "0000000000000000 A fixed_percpu_data",
+               "ffffffffb4231f40 D bpf_prog_fops",
+               "ffffffffb43723e0 d socket_file_ops",
+       }, "\n")
+
+       kAllSymsReader := strings.NewReader(kAllSymsStr)
+       kAllSyms, err := NewKAllSymsFromReader(kAllSymsReader)
+       require.NoError(t, err, "NewKAllSymsFromReader failed: %v", err)
+
        // Little endian representation of the addresses above:
        bpfProgFopsAddr := []byte{0x40, 0x1f, 0x23, 0xb4, 0xff, 0xff, 0xff, 0xff}
        socketFileOpsAddr := []byte{0xe0, 0x23, 0x37, 0xb4, 0xff, 0xff, 0xff, 0xff}
 
        spec := specTest.Copy()
 
-       err := specUpdateAddresses(
+       err = specUpdateAddresses(
                []symbolResolver{
-                       symbolResolverFromKallsyms(kAllSymsFactory),
+                       symbolResolverFromKallsyms(kAllSyms),
                },
                spec,
                []string{"abcde_bad_name"},
@@ -151,7 +151,7 @@ func TestSpecRewriteConstants(t *testing.T) {
 
        err = specUpdateAddresses(
                []symbolResolver{
-                       symbolResolverFromKallsyms(kAllSymsFactory),
+                       symbolResolverFromKallsyms(kAllSyms),
                },
                spec,
                []string{"bpf_prog_fops", "socket_file_ops"},
@@ -169,12 +169,15 @@ func TestSpecRewriteConstantsRoot(t *testing.T) {
        utilstest.RequireRoot(t)
        t.Cleanup(resetState)
 
+       kAllSyms, err := NewKAllSyms()
+       require.NoError(t, err, "NewKAllSyms failed: %v", err)
+
        resetState()
 
        spec1 := specTest.Copy()
-       err := specUpdateAddresses(
+       err = specUpdateAddresses(
                []symbolResolver{
-                       symbolResolverFromKallsyms(NewKAllSyms),
+                       symbolResolverFromKallsyms(kAllSyms),
                },
                spec1,
                []string{"bpf_prog_fops", "socket_file_ops"},
@@ -198,7 +201,7 @@ func TestSpecRewriteConstantsRoot(t *testing.T) {
        spec3 := specTest.Copy()
        err = specUpdateAddresses(
                []symbolResolver{
-                       symbolResolverFromKallsyms(NewKAllSyms),
+                       symbolResolverFromKallsyms(kAllSyms),
                        symbolResolverFromEbpf(),
                },
                spec3,

```